timeutil/schedule.go

```diff
@@ -0,0 +1,251 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package timeutil
+
+import (
+	"fmt"
+	"math/rand"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var validTime = regexp.MustCompile(`^([0-9]|0[0-9]|1[0-9]|2[0-3]):([0-5][0-9]):?([0-5][0-9])?$`)
+
+type TimeOfDay struct {
+	Hour   int
+	Minute int
+	Second int
+}
+
+func (td TimeOfDay) Less(other TimeOfDay) bool {
+	if td.Hour < other.Hour {
+		return true
+	}
+	if td.Minute < other.Minute {
+		return true
+	}
+	if td.Second < other.Second {
+		return true
+	}
+	return false
+}
+
+// ParseTime parses a string that contains hour:minute and returns
+// an TimeOfDay type or an error
+func ParseTime(s string) (t TimeOfDay, err error) {
+	m := validTime.FindStringSubmatch(s)
+	if len(m) == 0 {
+		return t, fmt.Errorf("cannot parse %q", s)
+	}
+	t.Hour, err = strconv.Atoi(m[1])
+	if err != nil {
+		return t, fmt.Errorf("cannot parse %q: %s", m[1], err)
+	}
+	t.Minute, err = strconv.Atoi(m[2])
+	if err != nil {
+		return t, fmt.Errorf("cannot parse %q: %s", m[2], err)
+	}
+	if m[3] != "" {
+		t.Second, err = strconv.Atoi(m[3])
+		if err != nil {
+			return t, fmt.Errorf("cannot parse %q: %s", m[3], err)
+		}
+	}
+	return t, nil
+}
+
+// Schedule defines a start and end time and an optional weekday in which
+// events should run.
+type Schedule struct {
+	Start TimeOfDay
+	End   TimeOfDay
+
+	Weekday string
+}
+
+func (sched *Schedule) Next(last time.Time) (start, end time.Time) {
+	now := timeNow()
+	wd := time.Weekday(weekdayMap[sched.Weekday])
+
+	t := last
+	for {
+		// slightly stupid, we move only a single day forward
+		t = t.AddDate(0, 0, 1)
```

Comment:
Subtle issues that I didn't realize they were subtle :/ Thanks for pointing me towards it again, I updated the code now. There was the issue of an endless loop in the playground when a weekly schedule was used and the update was outside of the 7 days (the last test case). But that can be fixed with a simple fix in the `if delta < 0` ->  `if delta <= 0`.